src/type-layout.md

```diff
@@ -56,12 +56,23 @@ Most primitives are generally aligned to their size, although this is
 platform-specific behavior. In particular, on x86 u64 and f64 are only
 aligned to 32 bits.
 
+For the primitive numeric types (`u8`, `i8`, `u16`, `i16`, `u32`, `i32`, `u64`,
+`i64`, `u128`, `i128`, `usize`, `isize`, `f32`, and `f64`), every bit pattern
+represents a valid instance of the type (in other words, for every primitive numeric
+type, `T`, `transmute::<[u8; size_of::<T>()], T>(...)` is always sound). For the
+primitive numeric types and also for `bool` and `char`, every byte is guaranteed to be
+initialized (in other words, for every such type, `T`,
+`transmute::<T, [u8; size_of::<T>()]>(...)` is always sound).
```

Comment:
Also the properties you are talking about here are not "layout" properties any more, they go deep into the operational semantics / value representation. Not sure if this is the right place in the reference to put them. (But then, not sure where to put anything about the op.sem into the reference.)

Reply:
> And (b) it presupposes that we allow pointer-to-integer transmutation, which is a subtle question and I am not sure if we're ready to commit to the answer yet.

I think pointers aren't covered by this paragraph since it says "primitive *numeric* type" which aren't part of the list above.

Reply:
Yeah we can definitely commit to transmuting between all the integer and float types.

Reply:
Is there accepted nomenclature for "all bit patterns that don't carry extra information in the abstract machine" or something to that effect? In other words, is there nomenclature for what your average programmer means when they say "all bit patterns"? I agree that the transmutes are precise, but IMO the prose is more intuitive for most readers.

Reply:
I'm not aware of a standard term for that.

> In other words, is there nomenclature for what your average programmer means when they say "all bit patterns"?

I've seen plenty of programmers include uninitialized memory in "all bit patterns". So I don't think we can make the assumption that they mean what you say they mean.

Reply:
Gotcha. How does this sound? (note the addition of "initialized" in the first sentence)

> For the primitive numeric types (`u8`, `i8`, `u16`, `i16`, `u32`, `i32`, `u64`,
`i64`, `u128`, `i128`, `usize`, `isize`, `f32`, and `f64`), every initialized bit pattern
represents a valid instance of the type (in other words, for every primitive numeric
type, `T`, `transmute::<[u8; size_of::<T>()], T>(...)` is always sound). Note that this
does not include uninitialized bit patterns, so transmuting from a type with padding
or other uninitialized bytes is unsound.

Reply:
"every initialized bit pattern represents a valid instance of the type" to me still sounds like "I can transmute pointers to the type".

Reply:
Yeah that makes sense. I don't have a good sense of what nomenclature is used for these different concepts; do you have a wording you'd suggest?

Reply:
I'm not aware of a standard term for "bytes that are initialized and don't have provenance". We could call them "pure bytes" or so, but even if a short term existed, it wouldn't be helpful -- people don't expect all this trouble with provenance, so there is no short term we can use to alert them to it.

I think for now this should just say the for every primitive numeric type, transmuting between them is always sound. And it should probably also explicitly say that the semantics of transmuting a pointer to an integer is undecided and that operation should be avoided. (That's what the `transmute` docs say.)

I don't think anyone ever doubted that transmuting between integer types was sound, so I am also somewhat wondering what the goal of this PR is. And is this really the right place to document transmutation rules? Wouldn't that be in the `transmute` docs?

Reply:
> I think for now this should just say the for every primitive numeric type, transmuting between them is always sound. And it should probably also explicitly say that the semantics of transmuting a pointer to an integer is undecided and that operation should be avoided. (That's what the `transmute` docs say.)

Gotcha; how does this sound?

> For every primitive numeric type (`u8`, `i8`, `u16`, `i16`, `u32`, `i32`, `u64`, `i64`, `u128`, `i128`, `usize`, `isize`, `f32`, and `f64`), `T`, the bit validity of `T` is equivalent to the bit validity of `[u8; size_of::<T>()]`. `u8` has 256 valid representations (namely, every 8-bit sequence). The following are not valid representations for `u8`:
> - A padding or uninitialized byte
> - A byte at any offset in a reference or pointer type (the semantics of transmuting a reference or pointer to a non-pointer type is currently undecided)

> I don't think anyone ever doubted that transmuting between integer types was sound, so I am also somewhat wondering what the goal of this PR is. Is this really the right place to document transmutation rules? Wouldn't that be in the `transmute` docs?

My specific use case is that I want to be able to write a safety comment on `unsafe impl FromBytes for T` (where `T` is one of the primitive types described in this paragraph) that proves the soundness of that impl. While I agree that everybody agrees that this is sound, I try to hold myself to a higher standard for safety comments - in particular, I want all of my safety comments to be proofs whose validity logically follows from text in the reference or the standard library documentation. My intention with [PRs](https://github.com/rust-lang/rust/pull/104081) [like](https://github.com/rust-lang/rust/pull/88915) [these](https://github.com/rust-lang/rust/pull/94786) is to codify this understanding so that even people who are completely unaware of the current Rust zeitgeist can still reason about the soundness of their code (namely, based purely on what's guaranteed by official docs).

Another reason that I call out `unsafe impl FromBytes` is that I don't actually think `transmute` is the right place for these things. At least in my mind, `transmute` is a function whose soundness is based on what facts are guaranteed about type layout and validity. There are ways to do what is effectively a transmute without using the `transmute` function itself. IMO the soundness of `transmute` ought to be downstream of these guarantees, and those guarantees should be documented in a central location. My reading is that, when you use `transmute` as an anchor in your soundness proofs, it's really shorthand for a more verbose argument like:
- `transmute::<X, Y>()` is documented as sound.
- Lemma: <some claim about Rust's layout and bit validity guarantees>
  - Proof by contradiction: If <some claim about Rust's layout and bit validity guarantees> were false, `transmute::<X, Y>()` would be unsound.
- Proof: Because <some claim about Rust's layout and bit validity guarantees> is true, what we're doing in this code is sound.

I recognize the irony of making that point on a PR in which I've used `transmute` in the Rust reference itself; it's just that I don't think we have good language for these things right now, and using `transmute` is an understood shorthand. My hope is that at least putting this in the reference (rather than just in the `transmute` docs) is a step in the right direction.

Reply:
> I'm not aware of a standard term for "bytes that are initialized and don't have provenance".

Perhaps "sequence of octets"? IIUC, this is what data formats developed before cchar stabilized on 8-bits call 'normal' bytes.

Reply:
> The following are not valid representations for u8:

We should make it clear that parts of this is undecided, I think. This sounds too definite for my taste.^^

> I want all of my safety comments to be proofs whose validity logically follows from text in the reference or the standard library documentation. 

I like it. :)

>  At least in my mind, transmute is a function whose soundness is based on what facts are guaranteed about type layout and validity. 

I fully agree. However, I certainly would not have thought to look in this place in the reference for such information. At the least, the library docs should then point to the reference docs -- for example [pages like these](https://doc.rust-lang.org/nightly/std/primitive.i32.html).

Reply:
> > The following are not valid representations for u8:
> 
> We should make it clear that parts of this is undecided, I think. This sounds too definite for my taste.^^

How about:

> A padding or uninitialized byte is not a valid `u8`. A byte at any offset in a reference or pointer type _may_ not be a valid `u8` (the semantics of transmuting a reference or pointer to a non-pointer type is currently undecided).

> > At least in my mind, transmute is a function whose soundness is based on what facts are guaranteed about type layout and validity.
> 
> I fully agree. However, I certainly would not have thought to look in this place in the reference for such information. And `transmute` is the established terminology for "re-interpreting memory from one type to another". At the least, the library docs should then point to the reference docs -- for example [pages like these](https://doc.rust-lang.org/nightly/std/primitive.i32.html), but also the docs for `transmute`.

Yeah, at some point when I have more time I'm hoping to propose a bigger overhaul of everything that includes a) putting all of these docs in a uniform place and putting links to those docs in appropriate places and, b) nailing down terminology like layout, bit validity, ABI, etc, and updating existing docs to use that terminology consistently. One of those never-the-most-important-thing-on-my-plate things, so it's never made it off the back burner, but I'm feeling very tempted to do it soon.

Reply:
> > I'm not aware of a standard term for "bytes that are initialized and don't have provenance".
> 
> Perhaps "sequence of octets"? IIUC, this is what data formats developed before cchar stabilized on 8-bits call 'normal' bytes.

Personally I think of this as a term used in contexts where bytes _can't_ have any extra semantic meaning (e.g., a network packet sent on the wire); not sure I've seen it used in the context of abstract machines.

Reply:
Slightly edited:

> An uninitialized byte is not a valid u8. A byte at any offset in a reference or pointer type may not be a valid u8 (the semantics of transmuting a reference or pointer to a non-pointer type is currently undecided).

"uninitialized byte" is part of the memory content. "padding byte" is a completely different concept, it appears in questions such as "does type X have a padding byte at offset N". But a byte that just sits in memory doesn't have a type and hence it makes no sense to ask whether it is a padding byte.

---

> nailing down terminology like layout, bit validity, ABI, etc, and updating existing docs to use that terminology consistently

Yeah, that makes sense. FWIW I think bit validity itself isn't even the core concept, the core concept is the representation relation -- which has more information than bit validity, for instance for integers if we do allow ptr2int transmutes the representation relation will say that the provenance of the bytes is ignored when computing the value representation of the integer. See [this](https://github.com/RalfJung/minirust/blob/master/spec/lang/representation.md) for an example of how that can be spelled out in full precision.

Reply:
> Slightly edited:
> 
> > An uninitialized byte is not a valid u8. A byte at any offset in a reference or pointer type may not be a valid u8 (the semantics of transmuting a reference or pointer to a non-pointer type is currently undecided).
> 
> "uninitialized byte" is part of the memory content. "padding byte" is a completely different concept, it appears in questions such as "does type X have a padding byte at offset N". But a byte that just sits in memory doesn't have a type and hence it makes no sense to ask whether it is a padding byte.

Okay, updated. How does the new text look?

--

> > nailing down terminology like layout, bit validity, ABI, etc, and updating existing docs to use that terminology consistently
> 
> Yeah, that makes sense. FWIW I think bit validity itself isn't even the core concept, the core concept is the representation relation -- which has more information than bit validity, for instance for integers if we do allow ptr2int transmutes the representation relation will say that the provenance of the bytes is ignored when computing the value representation of the integer. See [this](https://github.com/RalfJung/minirust/blob/master/spec/lang/representation.md) for an example of how that can be spelled out in full precision.

Cool, thanks for that link!

Reply:
Can you remove that part?

> `u8` has 256 valid representations (namely, every 8-bit sequence).

We didn't define "8-bit sequence", but this is too close for comfort to "every bit pattern". ;) Also, if a pointer bytes can be valid then u8 has infinitely many representations -- still only 256 abstract values, but infinitely many ways to represent each of them.

Reply:
Hmmm good point. Is there something else we can say instead of just removing it? Worst comes to worst, removing it is still fine - most code that needs to justify its soundness in terms of this paragraph is probably "bottoming out" at `u8` anyway, so defining everything in terms of `u8` and not fully nailing down `u8` itself is probably actually fine. But it'd be good to be able to make guarantees about what `u8` is instead of just what it isn't (uninitialized, pointer).

Reply:
Without having laid down the terminology of values and representation, it's tricky. We could say that every u8 represents a mathematical integer in `0..256`. But how high is the risk that people will skip the "represents" and get this wrong? Also, it seems really strange to say that only about u8. Unlike C, Rust doesn't have a magic 1-byte integer types. All our `uN` types represent a mathematical integer in `0..(2.pow(N))`.

I feel if we want to say something like this, we need some text somewhere about how each type has two sides: a set of abstract values it can represent, and a set of low-level representations each representing one such value (with potentially more than one representation per value). Then this text here can link to those terms and therefore make it clear what they mean. We can nail down the set of abstract values for the integer types, `bool`, `char`, but we cannot fully lay down their sets of valid representations -- but we *can* say that they are sufficiently compatible, in the sense that a valid representation of any of them is also a valid representation of `[u8; size_of::<T>()]` (and for the integer types, this is an if-and-only-if).

Reply:
Yeah, that sounds reasonable. I'll just remove the "all 256 8-bit sequences" text for now, and we can follow up with that at some point in the future.